src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,542 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Linq;
+
+namespace Microsoft.Experimental.Collections
+{
+    /// <summary>
+    /// DictionarySlim<TKey, TValue> is similar to Dictionary<TKey, TValue> but optimized in three ways:
+    /// 1) It allows access to the value by ref replacing the common TryGetValue and Add pattern.
+    /// 2) It does not store the hash code (assumes it is cheap to equate values).
+    /// 3) It does not accept an equality comparer(assumes Object.GetHashCode() and Object.Equals() or overridden implementation are cheap and sufficient).
+    /// </summary>
+    [DebuggerTypeProxy(typeof(Extensions.DictionarySlimDebugView<,>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public class DictionarySlim<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey, TValue>> where TKey : IEquatable<TKey>
+    {
+        // We want to initialize without allocating arrays. We also want to avoid null checks.
+        // Array.Empty would give divide by zero in modulo operation. So we use static one element arrays.
+        // The first add will cause a resize replacing these with real arrays of three elements.
+        // Arrays are wrapped in a class to avoid being duplicated for each <TKey, TValue>
+        private static readonly int[] InitialBuckets = DummySizeOneArray<int>.Value;
```

Comment:
@jkotas is this nested class approach the correct way to avoid instantiation of these dummy arrays (which are not parameterized over `<TKey, TValue>` for every instantiation of the ` DictionarySlim<TKey, TValue>` type?

Reply:
> // Arrays are wrapped in a class to avoid being duplicated for each <TKey, TValue>

As implemented, this arrays are still duplicate for each `<TKey, TValue>`. You are just paying several hundred bytes extra for `DummySizeOneArray` instantiations, so this is deoptimization.

For InitialBuckets, you can move it to HashHelpers to share it.
For InitialEntries, just make it `private static readonly Entry[] InitialEntries = new Entry[1];`

Reply:
done thanks